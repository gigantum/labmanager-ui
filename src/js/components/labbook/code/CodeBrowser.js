// vendor
import React, { Component } from 'react'
import {createPaginationContainer, graphql} from 'react-relay'
//mutations
import FileBrowserWrapper from 'Components/labbook/fileBrowser/FileBrowserWrapper'
//store
import store from 'JS/redux/store'

class CodeBrowser extends Component {
  constructor(props){
  	super(props);

    this.state = {
      rootFolder: '',
      moreLoading: false
    }

    this.setRootFolder = this.setRootFolder.bind(this)
    this._loadMore = this._loadMore.bind(this)
  }

  /*
    loads more if branches are switched
  */
  componentDidUpdate(){
    if(!this.state.moreLoading && this.props.code.allFiles && this.props.code.allFiles.edges.length < 3 && this.props.code.allFiles.pageInfo.hasNextPage){
      this._loadMore();
    }
  }

  /*
    handle state and addd listeners when component mounts
  */
  componentDidMount() {
    if(this.props.code.allFiles &&
      this.props.code.allFiles.pageInfo.hasNextPage) {
        this._loadMore() //routes query only loads 2, call loadMore
    } else {
      this.setState({'moreLoading': false});
    }
  }
  /*
    @param
    triggers relay pagination function loadMore
    increments by 100
    logs callback
  */
  _loadMore() {
    this.setState({'moreLoading': true});
    let self = this;
    this.props.relay.loadMore(
     100, // Fetch the next 100 feed items
     (response, error) => {
       if(error){
         console.error(error)
       }

       if(self.props.code.allFiles &&
         self.props.code.allFiles.pageInfo.hasNextPage) {
         self._loadMore()
       } else {
         this.setState({'moreLoading': false});
      }
     }
   );
  }
  /*
    @param
    sets root folder by key
    loads more files
  */
  setRootFolder(key){
    this.setState({rootFolder: key})
  }

  render(){

    this.props.loadStatus(this.state.moreLoading);
    if(this.props.code && this.props.code.allFiles){

      let codeFiles = this.props.code.allFiles
      if(this.props.code.allFiles.edges.length === 0){
        codeFiles = {
          edges: [],
          pageInfo: this.props.code.allFiles.pageInfo
        }
      }
      return(
          <FileBrowserWrapper
            ref='codeBrowser'
            section="code"
            selectedFiles={this.props.selectedFiles}
            clearSelectedFiles={this.props.clearSelectedFiles}
            setRootFolder={this.setRootFolder}
            files={codeFiles}
            parentId={this.props.codeId}
            connection="CodeBrowser_allFiles"
            favoriteConnection="CodeFavorites_favorites"
            favorites={this.props.favorites}
            isLocked={this.props.isLocked}
            {...this.props}
          />
      )
    }else{
      return(<div>No Files Found</div>)
    }
  }
}

export default createPaginationContainer(
  CodeBrowser,
  {

    code: graphql`
      fragment CodeBrowser_code on LabbookSection{
        allFiles(after: $cursor, first: $first)@connection(key: "CodeBrowser_allFiles", filters: []){
          edges{
            node{
              id
              isDir
              isFavorite
              modifiedAt
              key
              size
            }
            cursor
          }
          pageInfo{
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
          }
        }

      }`
  },
  {
    direction: 'forward',
    getConnectionFromProps(props) {
      return props.code && props.code.allFiles
    },
    getFragmentVariables(prevVars, totalCount,) {

      return {
        ...prevVars,
        first: totalCount,
      };
    },
    getVariables(props, {count, cursor}, fragmentVariables) {
      const {owner, labbookName} = store.getState().routes

      return {
        first: count,
        cursor: cursor,
        owner: owner,
        name: labbookName
      };
    },
    query: graphql`
      query CodeBrowserPaginationQuery(
        $first: Int
        $cursor: String
        $owner: String!
        $name: String!
      ) {
        labbook(name: $name, owner: $owner){
           id
           description
           code{
             id
            # You could reference the fragment defined previously.
            ...CodeBrowser_code

          }
        }
      }
    `
  }

)
