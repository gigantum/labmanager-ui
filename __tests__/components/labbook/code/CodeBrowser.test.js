
      import React, {Component} from 'react'
      import renderer from 'react-test-renderer';
      import {mount} from 'enzyme'
      import CodeBrowser from 'Components/labbook/code/CodeBrowser';

      import { DragDropContext } from 'react-dnd'
      import HTML5Backend from 'react-dnd-html5-backend'

      import json from './__relaydata__/CodeBrowser.json'

      import relayTestingUtils from 'relay-testing-utils'

      const loadStatus = ()=>{

      }


      const clearSelectedFiles = ()=>{

      }

      const backend = (manager: Object) => {
          const backend = HTML5Backend(manager),
              orgTopDropCapture = backend.handleTopDropCapture;

          backend.handleTopDropCapture = (e) => {

              if(backend.currentNativeSource){
                orgTopDropCapture.call(backend, e);

               //backend.currentNativeSource.item.dirContent = getFilesFromDragEvent(e, {recursive: true}); //returns a promise
              }
          };

          return backend;
      }

      const fixtures = {
        labbook: json.data.labbook,
        labbookId: json.data.labbook.id,
        isLocked: false,
        selectedFiles: [],
        clearSelectedFiles,
        codeId: json.data.labbook.code.id,
        code: json.data.labbook.code,
        loadStatus
      }

      class CodeCompInstance extends Component{
        render(){
          return (relayTestingUtils.relayWrap(<CodeBrowser {...fixtures}/>, {}, json.data.labbook))
        }
      }
      const CodeBrowserComponent = DragDropContext(backend)(CodeCompInstance)

      describe('Test CodeBrowser', () => {

        it('snapshot renders', () => {
          const wrapper = renderer.create(
             <CodeBrowserComponent />
          );

          const tree = wrapper.toJSON()

          expect(tree).toMatchSnapshot()
        })


      })
