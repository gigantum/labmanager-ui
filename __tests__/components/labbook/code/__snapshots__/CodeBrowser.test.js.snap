// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CodeBrowser snapshot renders 1`] = `
<div
  className="Code flex flex-row justify-center"
  id="code"
>
  <div
    className="rendered-react-keyed-file-browser"
  >
    <div
      className="rendered-file-browser"
    >
      <div
        className="action-bar"
      >
        <input
          onChange={[Function]}
          placeholder="Filter files"
          type="search"
          value=""
        />
        <ul
          className="item-actions"
        >
          <li>
            <a
              className="btn btn-primary btn-sm"
              href="#"
              onClick={[Function]}
              role="button"
            >
              <i
                aria-hidden="true"
                className="fa fa-folder-o"
              />
              Â Add Folder
            </a>
          </li>
        </ul>
      </div>
      <div
        className="files"
        id="CodeBrowser_allFiles"
      >
        <table
          cellPadding="0"
          cellSpacing="0"
        >
          <thead>
            <tr
              className="folder"
            >
              <th>
                File
              </th>
              <th />
              <th
                className="size"
              >
                Size
              </th>
              <th
                className="modified"
              >
                Last Modified
              </th>
              <th />
            </tr>
          </thead>
          <tbody>
            <tr
              className="folder"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="fa fa-folder-o"
                      />
                      @1xGigantum (1)
                    </a>
                  </div>
                </div>
              </td>
              <td />
              <td />
              <td />
            </tr>
            <tr
              className="folder"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      onClick={[Function]}
                    >
                      <i
                        aria-hidden="true"
                        className="fa fa-folder-o"
                      />
                      export
                    </a>
                  </div>
                </div>
              </td>
              <td />
              <td />
              <td />
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon database-icon"
                      />
                      datasets.yml
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                4 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon svg-icon"
                      />
                      down-arrow.svg
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                463 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon null"
                      />
                      down-arrow.textClipping
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                203 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon database-icon"
                      />
                      environment.yml
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                744 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon python-icon"
                      />
                      filetimes.py
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                16 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon terminal-icon"
                      />
                      filetimes.sh
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                3 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon database-icon"
                      />
                      filetimes.yml
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                521 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon terminal-icon"
                      />
                      get_raster_data.sh
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                138 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon jupyter-icon"
                      />
                      nyc_taxi-nongeo.ipynb
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                8.70 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon jupyter-icon"
                      />
                      nyc_taxi.ipynb
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                11.34 MB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon python-icon"
                      />
                      pcap_to_parquet.py
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                3 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon python-icon"
                      />
                      raster.py
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                3 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon book-icon"
                      />
                      README.md
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                8 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon database-icon"
                      />
                      small.yml
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                484 B
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon jupyter-icon"
                      />
                      streaming-aggregation.ipynb
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                15 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon python-icon"
                      />
                      streaming.py
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                5 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
            <tr
              className="file modified"
              onClick={[Function]}
              onDoubleClick={[Function]}
            >
              <td
                className="name"
              >
                <div
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                >
                  <div>
                    <a
                      download="download"
                      href="#"
                      onClick={[Function]}
                    >
                      <span
                        className="FileBrowser__icon python-icon"
                      />
                      taxi_preprocessing_example.py
                    </a>
                  </div>
                </div>
              </td>
              <td
                width="30"
              >
                <div
                  className="Favorite__star--empty"
                  onClick={[Function]}
                />
              </td>
              <td
                className="size"
              >
                3 kB
              </td>
              <td
                className="modified"
              >
                2 months ago
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
  <div
    className="DetailPanel hidden"
  >
    <div
      className="DetailPanel--close"
      onClick={[Function]}
    >
      X
    </div>
    <p />
    <p />
  </div>
</div>
`;
